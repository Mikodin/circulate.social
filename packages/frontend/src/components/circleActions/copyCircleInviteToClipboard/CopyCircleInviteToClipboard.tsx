import { useState } from 'react';
import * as copy from 'copy-to-clipboard';
import { Input, Button, Modal } from 'antd';
import {
  CheckCircleOutlined,
  UserAddOutlined,
  CopyOutlined,
} from '@ant-design/icons';

import styles from './copyCircleInviteToClipboard.module.scss';

export interface Props {
  circleId: string;
  circleName: string;
}
const domain = 'beta.circulate.social';

const CopyCircleInviteToClipboard = (props: Props): JSX.Element => {
  const { circleId, circleName } = props;
  const baseInviteText = `Hey, I'd love for you to join our Circle "${circleName}".\n\nhttps://${domain}/circles/${circleId}?join=true`;

  const [isModalOpen, setIsModalOpen] = useState(false);
  const [isTextCopied, setIsTextCopied] = useState(false);
  const [inviteText, setInviteText] = useState(baseInviteText);

  const copyText = () => {
    copy.default(inviteText);
    setIsTextCopied(true);
    setTimeout(() => {
      setIsTextCopied(false);
      setIsModalOpen(false);
    }, 2000);
  };
  const copyToClipboardButton = (
    <Button
      key="inviteButton"
      size="middle"
      onClick={copyText}
      type={isTextCopied ? 'primary' : 'default'}
      icon={isTextCopied ? <CheckCircleOutlined /> : <CopyOutlined />}
    >
      {isTextCopied ? 'Copied!' : 'Copy To Clipboard'}
    </Button>
  );

  return (
    <div>
      <Button
        onClick={() => {
          setIsModalOpen(true);
        }}
        icon={<UserAddOutlined />}
      >
        Invite
      </Button>

      <Modal
        className={styles.modal}
        visible={isModalOpen}
        title="Invite someone"
        onOk={() => setIsModalOpen(false)}
        onCancel={() => {
          setIsModalOpen(false);
          setInviteText(baseInviteText);
        }}
        footer={[copyToClipboardButton]}
      >
        <Input.TextArea
          rows={4}
          value={inviteText}
          onChange={(event) => {
            setInviteText(event.target.value);
          }}
        />
      </Modal>
    </div>
  );
};

export default CopyCircleInviteToClipboard;
