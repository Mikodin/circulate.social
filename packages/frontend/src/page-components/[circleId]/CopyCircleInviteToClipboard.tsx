import { useState } from 'react';
import * as copy from 'copy-to-clipboard';
import { Input, Button } from 'antd';
import { CopyOutlined, CheckCircleOutlined } from '@ant-design/icons';

export interface Props {
  circleId: string;
}
const domain = 'beta.circulate.social';

const CopyCircleInviteToClipboard = (props: Props): JSX.Element => {
  const { circleId } = props;
  const [isTextCopied, setIsTextCopied] = useState(false);
  return (
    <Input
      size="small"
      addonBefore={
        <Button
          size="small"
          onClick={() => {
            copy.default(`https://${domain}/circles/${circleId}?join=true`);
            setIsTextCopied(true);

            setTimeout(() => {
              setIsTextCopied(false);
            }, 2000);
          }}
          type={isTextCopied ? 'primary' : 'default'}
          icon={isTextCopied ? <CheckCircleOutlined /> : <CopyOutlined />}
          block
        >
          {isTextCopied ? 'Copied!' : 'Invite'}
        </Button>
      }
      readOnly
      value={`https://${domain}/circles/${circleId}?join=true`}
    />
  );
};

export default CopyCircleInviteToClipboard;
