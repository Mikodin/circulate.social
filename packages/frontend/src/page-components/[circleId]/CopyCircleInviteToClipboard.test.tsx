import {
  render,
  RenderResult,
  act,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import * as copy from 'copy-to-clipboard';
import CopyCircleInviteToClipboard, {
  Props,
} from './CopyCircleInviteToClipboard';

const DEFAULT_PROPS = {
  circleId: 'asdf-123',
};
function renderCopyCircleInviteToClipboard(
  props?: Partial<Props>
): RenderResult {
  return render(<CopyCircleInviteToClipboard {...DEFAULT_PROPS} {...props} />);
}

describe('CopyCircleInviteToClipboard', () => {
  it('Should render', () => {
    expect(renderCopyCircleInviteToClipboard()).toMatchSnapshot();
  });

  it('Should call copy from copy-to-clipboard library on button click', () => {
    const copySpy = jest.spyOn(copy, 'default');
    const { queryByText } = renderCopyCircleInviteToClipboard();

    const copyButton = queryByText(/Invite/i);
    act(() => {
      fireEvent.click(copyButton);
    });

    expect(copySpy).toHaveBeenCalled();
  });

  it('Should render "Copied!" on button click', () => {
    const { queryByText } = renderCopyCircleInviteToClipboard();

    const copyButton = queryByText(/Invite/i);
    act(() => {
      fireEvent.click(copyButton);
    });

    waitFor(() => {
      expect(queryByText(/Copied!/i)).toBeTruthy();
    });
  });
});
