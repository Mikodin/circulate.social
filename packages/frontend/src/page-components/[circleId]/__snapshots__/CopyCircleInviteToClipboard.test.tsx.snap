// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyCircleInviteToClipboard Should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="ant-input-group-wrapper ant-input-group-wrapper-sm"
      >
        <span
          class="ant-input-wrapper ant-input-group"
        >
          <span
            class="ant-input-group-addon"
          >
            <button
              class="ant-btn ant-btn-default ant-btn-sm ant-btn-block"
              type="button"
            >
              <span
                aria-label="copy"
                class="anticon anticon-copy"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="copy"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                  />
                </svg>
              </span>
              <span>
                Invite
              </span>
            </button>
          </span>
          <input
            class="ant-input ant-input-sm"
            readonly=""
            type="text"
            value="https://beta.circulate.social/circles/asdf-123?join=true"
          />
        </span>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="ant-input-group-wrapper ant-input-group-wrapper-sm"
    >
      <span
        class="ant-input-wrapper ant-input-group"
      >
        <span
          class="ant-input-group-addon"
        >
          <button
            class="ant-btn ant-btn-default ant-btn-sm ant-btn-block"
            type="button"
          >
            <span
              aria-label="copy"
              class="anticon anticon-copy"
              role="img"
            >
              <svg
                aria-hidden="true"
                class=""
                data-icon="copy"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                />
              </svg>
            </span>
            <span>
              Invite
            </span>
          </button>
        </span>
        <input
          class="ant-input ant-input-sm"
          readonly=""
          type="text"
          value="https://beta.circulate.social/circles/asdf-123?join=true"
        />
      </span>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
